File: src/main/java/one/oktw/sponge/internal/galaxy/PlanetManager.java
package one.oktw.sponge.internal.galaxy;

import com.mongodb.client.MongoCollection;
import one.oktw.sponge.Main;
import org.bson.Document;
import org.slf4j.Logger;
import org.spongepowered.api.Server;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.world.World;
import org.spongepowered.api.world.WorldArchetypes;
import org.spongepowered.api.world.storage.WorldProperties;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ExecutionException;

import static com.mongodb.client.model.Filters.eq;
import static one.oktw.sponge.Main.getMain;
import static one.oktw.sponge.internal.galaxy.SecurityLevel.VISIT;

public class PlanetManager {
    private Main main = getMain();
    private Logger logger = main.getLogger();
    private MongoCollection<Document> collection = main.getDatabaseManager().getDatabase().getCollection("World");
    private Server server = Sponge.getServer();

    Planet createWorld(String name) {
        WorldProperties properties;
        logger.info("Create World [{}]", name);

        try {
            properties = server.createWorldProperties(name, WorldArchetypes.OVERWORLD);
            properties.setKeepSpawnLoaded(false);
            properties.setGenerateSpawnOnLoad(false);
            properties.setLoadOnStartup(false);
            server.saveWorldProperties(properties);
        } catch (IOException e) {
            logger.error("Create world failed!", e);
            throw new UncheckedIOException(e);
        }

        Document worldInfo = new Document("UUID", properties.getUniqueId())
                .append("Name", name)
                .append("Size", 32)
                .append("Security", VISIT);

        collection.insertOne(worldInfo);
        return new Planet(properties.getUniqueId());
    }

    public void removeWorld(UUID uuid) {
        WorldProperties properties;
        if (server.getWorldProperties(uuid).isPresent()) {
            properties = server.getWorldProperties(uuid).get();
        } else {
            logger.error("Delete World [{}] failed: world not found", uuid.toString());
            return;
        }

        logger.info("Deleting World [{}]", properties.getWorldName());
        if (server.getWorld(uuid).isPresent()) {
            World world = server.getWorld(uuid).get();
            world.getPlayers().forEach(player -> player.setLocationSafely(server.getWorld(server.getDefaultWorldName()).get().getSpawnLocation()));
            server.unloadWorld(world);
        }

        try {
            server.deleteWorld(properties).get();
        } catch (InterruptedException | ExecutionException e) {
            logger.error("Delete world failed!", e);
            return;
        }

        collection.deleteOne(eq("UUID", uuid));
    }

    Optional<World> loadWorld(UUID uuid) {
        if (server.getWorldProperties(uuid).isPresent()) {
            WorldProperties worldProperties = server.getWorldProperties(uuid).get();
            worldProperties.setGenerateSpawnOnLoad(false);
            return server.loadWorld(worldProperties);
        } else {
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/one/oktw/galaxy/internal/PlanetManager.kt
package one.oktw.galaxy.internal

import one.oktw.galaxy.Main.Companion.main
import one.oktw.galaxy.internal.types.Planet
import org.spongepowered.api.Sponge
import org.spongepowered.api.world.World
import org.spongepowered.api.world.WorldArchetypes
import org.spongepowered.api.world.storage.WorldProperties
import java.io.IOException
import java.io.UncheckedIOException
import java.util.*
import java.util.concurrent.CompletableFuture

internal class PlanetManager {
    companion object {
        private val logger = main.logger
        private val server = Sponge.getServer()

        fun createPlanet(name: String): Planet {
            val properties: WorldProperties
            logger.info("Create World [{}]", name)

            try {
                properties = server.createWorldProperties(name, WorldArchetypes.OVERWORLD)
                properties.setKeepSpawnLoaded(false)
                properties.setGenerateSpawnOnLoad(false)
                properties.setLoadOnStartup(false)
                server.saveWorldProperties(properties)
            } catch (e: IOException) {
                logger.error("Create world failed!", e)
                throw UncheckedIOException(e)
            }

            val planet = Planet(world = properties.uniqueId, name = name)
            return planet
        }

        fun removePlanet(worldUUID: UUID): CompletableFuture<Boolean>? {
            val properties: WorldProperties
            if (server.getWorldProperties(worldUUID).isPresent) {
                properties = server.getWorldProperties(worldUUID).get()
            } else {
                logger.error("Delete World [{}] failed: world not found", worldUUID.toString())
                return null
            }

            logger.info("Deleting World [{}]", properties.worldName)
            if (server.getWorld(worldUUID).isPresent) {
                val world = server.getWorld(worldUUID).get()
                world.players.parallelStream().forEach { player -> player.setLocationSafely(server.getWorld(server.defaultWorldName).get().spawnLocation) }
                server.unloadWorld(world)
            }

            return server.deleteWorld(properties)
        }

        fun loadPlanet(uuid: UUID): Optional<World> {
            return if (server.getWorldProperties(uuid).isPresent) {
                val worldProperties = server.getWorldProperties(uuid).get()
                worldProperties.setGenerateSpawnOnLoad(false)
                server.loadWorld(worldProperties)
            } else {
                Optional.empty()
            }
        }
    }
}
